File: 2.两数相加.cpp
/*
 * @lc app=leetcode.cn id=2 lang=cpp
 *
 * [2] 两数相加
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        if (l1 == NULL)
        {
            return l2;
        }
        if (l2 == NULL)
        {
            return l1;
        }

        int carry_value = 0;
        ListNode* tail = nullptr;
        ListNode* head = nullptr;
        while (l1 || l2)
        {
            int l1_val = (l1 == NULL) ? 0 : l1->val;
            int l2_val = (l2 == NULL) ? 0 : l2->val;

            int val = l1_val + l2_val + carry_value;
            carry_value = val / 10;
            
            if (head)
            {
                tail->next = new ListNode(val % 10);
                tail = tail->next;
            }
            else
            {
                head = new ListNode(val % 10);
                tail = head;
            }
            if (l1)
            l1 = l1->next;
            if (l2)
            l2 = l2->next;
        }

        if (carry_value != 0)
        {
            tail->next = new ListNode(carry_value);
            tail = tail->next;
        }

        return head;
    }
};
// @lc code=end

File: 2.两数相加.go
/*
 * @lc app=leetcode.cn id=2 lang=golang
 *
 * [2] 两数相加
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var head, tail *ListNode
	carry_val := 0
	for l1 != nil || l2 != nil{
		l1_val, l2_val := 0, 0

		if l1 != nil {
			l1_val = l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			l2_val = l2.Val
			l2 = l2.Next
		}

		sum := l1_val + l2_val + carry_val
		sum, carry_val = sum%10, sum/10

		if head == nil {
			head = &ListNode{Val: sum}
			tail = head
		} else {
			tail.Next = &ListNode{Val: sum}
			tail = tail.Next
		}
	}

	if carry_val != 0 {
		tail.Next = &ListNode{Val: carry_val}
		tail = tail.Next
	}
	return head
}
// @lc code=end

